Fixed required for running qa_test_klp in kernel v5.3

lpechacek opened the pull request:
Recent kernel API changes affected KLP tests. They are HR timer changes, decomissioning of jbprobes and syscall entry reimplementation on x86_64. Detect these changes at run time and reflect them in test behavior.

Commit:
Fix test module build

With v5.3 test kmod build fails with

  [07:42:51]  ERROR: Kernel configuration is invalid.
  [07:42:51]          include/generated/autoconf.h or
include/config/auto.conf are missing.
  [07:42:51]          Run 'make oldconfig && make prepare' on kernel
src to fix it.
  [07:42:51]  /usr/src/linux-5.3.8-2/Makefile:613:
include/config/auto.conf: No such file or directory

The cause lies in changed behavior of KBUILD_OUTPUT (AKA "O=") which now
makes kbuild search for valid configuration in the output directory.
Given that we expect the output files to go into the same place where
the sources are stored, setting M is enough.

Signed-off-by: Libor Pecháček <lpechacek@suse.cz>

Commit:
Drop support for Jprobe

This was unused in the tests after all.

Signed-off-by: Libor Pecháček <lpechacek@suse.cz>

nicstange commented on klp_configure.sh:
```diff
@@ -0,0 +1,37 @@
+
+set -e
+. "$(dirname $0)"/klp_tc_functions.sh
+
```
klp_configure.sh gets sourced from klp_tc_functions.sh. Thus, klp_tc_functions.sh is effectively sourcing itself (indirectly) again which is a bit confusing. IMO, it would be cleaner to move the sourcing snippet for klp_configure.sh to the end of klp_tc_functions.sh (or right after the definition of klp_compile_module(), which it needs) and make the former not source klp_tc_functions.sh.

lpechacek replied:
It looks to me like the best course of action will be squashing `klp_configure.sh` into `klp_tc_functions.sh`. Then we will eliminate the include maze.

nicstange commented on klp_configure.sh:
```diff
@@ -0,0 +1,37 @@
+
+# Check for getpid syscall prefix
+echo -n 'export KLP_TEST_SYSCALL_FN_PREFIX='
+sed '/T [[:alnum:]_]*sys_getpid$/{ s/.* \([^ ]*\)sys_getpid$/\1/;p }; d' /proc/kallsyms | tail -1
```
If I'm not mistaken, there are two getpid() syscall stubs on x86_64 (starting with 4.17.0): __x64_sys_getpid() and __ia32_sys_getpid(). Likewise, on s390, there should be __s390x_sys_getpid() and __s390_sys_getpid(). Are you sure the above always catches the right variant? If not, we should implement some logic based on the kernel version and arch, similar to what is done in kgraft-patches/syscalls.h (master-livepatch branch) for the KLP_SYMCALL_SYM define.

lpechacek replied:
Good point. I happily took the last match in the hope that it will be the right one. That's overly fragile though. I'll glance at `kgraft-patches/syscalls.h`.

As an alternative, I thought about searching for `__$(uname -m)_sys_getpid` in kallsyms, and if it's not present, resort to plain `sys_getpid` (with possible test abort in case that even this symbol is missing). I'll weight the alternatives after studying your proposal.

lpechacek replied:
OK, the search for `__$(uname -m)_sys_getpid` didn't work. The thing is that on `x86_64` the syscall function name is `__x64_sys_getpid`. I'll rewrite the [kgraft-patches checks](https://kernel.suse.com/cgit/kernel-livepatch/tree/klp_syscalls.h?h=SLE15-SP2_Update_0) to shell then.

lpechacek replied:
So I've implemented a kernel version comparison and created a switch based on `uname -m`. Hope that's a better implementation.

nicstange commented on klp_configure.sh:
```diff
@@ -0,0 +1,37 @@
+cp "$SOURCE_DIR"/klp_compile_test_hrtimer.c $COMPILETEST_DIR/
+
+echo -n 'export KLP_TEST_HRTIMER_OLD='
+if klp_compile_module /tmp/live-patch/klp_compile_test/klp_compile_test_hrtimer.c > /dev/null 2>&1;
```
$COMPILETEST_DIR could be used in the source file specification.

lpechacek replied:
Yes, that one slipped through. Fixed. Thanks!

nicstange reviewed (commented):
Overall it's great. I'm only concerned about the KLP_TEST_SYSCALL_FN_PREFIX determination, which I believe could possibly yield the undesired 32bit variant. Other than that, two minor nits in the inline comments.

Commit:
Prepare for run-time kernel API checking

Recent kernel API changes affected KLP tests. Prepare infrastructure for
detection of these changes at run time and reflect them in environment
settings.

Signed-off-by: Libor Pecháček <lpechacek@suse.cz>

Commit:
Deal with HR timer API changes

Adapt test code to support both HR timer API variants.

Signed-off-by: Libor Pecháček <lpechacek@suse.cz>

lpechacek commented:
Thanks for the review, Nicolai. Here come updated patches. Despite the wrong indentation which sneaked into b2f5da0, I consider them ready for another review round. I'll clean up the white space mess for final merge. At this point I'd like to keep the number of branch updates low.
Comments and suggestions are welcome.

nicstange commented:
Looks good to me. The only thing I'd do before the merge is to replace the -gt comparison of VERSION_CODE by a -ge one: The syscalls got reworked in kernel 4.17. already.

Commit:
Deal with syscall function prefixes

Since v4.17 functions on certain architectures carry an
arch-specific prefix. Parametrize test kernel module with this prefix.

Signed-off-by: Libor Pecháček <lpechacek@suse.cz>

lpechacek commented:
Thanks for the thorough review, Nicolai! I've adjusted the kernel version test and will merge the changes.
